internal/api/handler.go

```diff
@@ -84,57 +103,141 @@ func (h *Handler) handleAuthorize(c *Context, w http.ResponseWriter, r *http.Req
+func (h *Handler) handleAuthorizationCallback(c *OAuthAppContext, w http.ResponseWriter, r *http.Request) {
+	c.Logger = loggerWithAppFields(c.Logger, c.OAuthApplication)
 	c.Logger.Info("Handling authorization callback")
 
 	state := r.URL.Query().Get("state")
 
-	redirectURIRaw, err := h.stateCache.GetRedirectURI(state)
+	chimeraAuthZState, err := h.getAuthZStateFromCache(state)
 	if err != nil {
-		if err == cache.ErrNotFound {
-			c.Logger.Error("State provided to webhook handler not found")
-			w.WriteHeader(http.StatusBadRequest)
-		} else {
-			c.Logger.Error("Failed to get state from cache")
-			w.WriteHeader(http.StatusInternalServerError)
-		}
+		c.Logger.WithError(err).Error("Failed to get redirect URI from cache")
+		w.WriteHeader(statuserr.ErrToStatus(err))
 		return
 	}
 
-	redirectURI, err := url.Parse(redirectURIRaw)
+	redirectURI, err := url.Parse(chimeraAuthZState.RedirectURI)
 	if err != nil {
 		c.Logger.Error("Failed to parse destination site URL")
 		w.WriteHeader(http.StatusInternalServerError)
 		return
 	}
 
+	// Update redirect URI with params returned from OAuth provider such as `authorization_code`.
 	originalState := state[:len(state)-stateExtensionLength]
-
 	query := r.URL.Query()
 	query.Set("state", originalState)
-
 	redirectURI.RawQuery = query.Encode()
-	c.Logger.Info("Redirecting authorization callback")
 
-	http.Redirect(w, r, redirectURI.String(), http.StatusFound)
+	redirectURIStr := redirectURI.String()
+
+	verificationToken := util.NewRandomString(32)
+	err = h.stateCache.SetRedirectURI(state, cache.AuthorizationState{RedirectURI: redirectURIStr, AuthorizationVerificationToken: verificationToken})
+	if err != nil {
+		c.Logger.Error("Failed to update redirect URL in cache")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	c.Logger.Info("Redirecting authorization callback to confirmation")
+	http.Redirect(w, r, fmt.Sprintf("%s/v1/%s/%s/auth/chimera/confirm?state=%s", h.baseURL, c.OAuthApplication.Provider, c.OAuthApplication.Identifier, state), http.StatusFound)
 }
 
-func (h *Handler) handleTokenExchange(c *Context, w http.ResponseWriter, r *http.Request) {
-	app, found := h.getOAuthApp(c, r)
-	if !found {
-		w.WriteHeader(http.StatusNotFound)
+func (h *Handler) handleGetConfirmAuthorization(c *OAuthAppContext, w http.ResponseWriter, r *http.Request) {
+	c.Logger = loggerWithAppFields(c.Logger, c.OAuthApplication)
+	c.Logger.Info("Handling request for confirmation of Chimera authZ")
+
+	state := r.URL.Query().Get("state")
+	fmt.Println("Handle confirm state: ", state)
+
+	chimeraAuthZState, err := h.getAuthZStateFromCache(state)
+	if err != nil {
+		c.Logger.WithError(err).Error("Failed to get redirect URI from cache")
+		w.WriteHeader(statuserr.ErrToStatus(err))
 		return
 	}
-	c.Logger = c.Logger.WithField("provider", app.Provider).
-		WithField("app", app.Identifier)
+
+	// Strip URL query for cleaner display
+	strippedURL, err := stripURLQuery(chimeraAuthZState.RedirectURI)
```

Comment:
Like I suspected earlier, more validation is needed here before the URL can be displayed cleanly. I'll DM you with details.